File: leetcode/2-add-two-numbers.cpp
/*
You are given two linked lists representing two non-negative numbers. The digits are stored in reverse order and each of their nodes contain a single digit. Add the two numbers and return it as a linked list.

Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
Output: 7 -> 0 -> 8
*/
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        auto carry(0);
        ListNode dummy(0);
        auto head(&dummy);
        while (l1 != nullptr && l2 != nullptr) {
            auto result(l1->val + l2->val + carry);
            carry = result / 10;
            head->next = new ListNode(result % 10);
            head = head->next;
            l1 = l1->next;
            l2 = l2->next;
        }
        auto x(l1 == nullptr ? l2 : l1);
        while (x != nullptr) {
            auto result(x->val + carry);
            carry = result / 10;
            head->next = new ListNode(result % 10);
            head = head->next;
            x = x->next;
        }
        if (carry != 0) {
            head->next = new ListNode(carry);
            head = head->next;
        }
        return dummy.next;
    }
};

File: leetcode/2-add-two-numbers.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} l1
# @param {ListNode} l2
# @return {ListNode}
def add_two_numbers(l1, l2)
  carry = 0
  dummy = ListNode.new 0
  current = dummy
  while (not l1.nil?) and (not l2.nil?)
    tmp = carry + l1.val + l2.val
    carry = tmp / 10
    tmp = tmp % 10
    node = ListNode.new tmp
    current.next = node
    current = current.next
    l1 = l1.next
    l2 = l2.next
  end

  l = l1.nil? ? l2 : l1
  while not l.nil?
    tmp = carry + l.val
    carry = tmp / 10
    tmp = tmp % 10
    node = ListNode.new tmp
    current.next = node
    current = current.next
    l = l.next
  end

  while carry > 0
    node = ListNode.new carry % 10
    current.next = node
    current = current.next
    carry = carry / 10
  end

  dummy.next
end
